rfd/xxxx-gcp-server-discovery.md

```diff
@@ -0,0 +1,285 @@
+state: draft
+---
+
+# RFD X - GCP auto-discovery
+
+## Required Approvers
+
+Engineering: @jakule && @r0mant
+Product: @klizhentas && @xinding33
+Security: @reedloden
+
+## What
+
+Teleport discovery services will be able to automatically discover and enroll GCP virtual machine
+instances.
+
+## Why
+
+This change will bring GCP auto-joining capabilities in line with EC2 and Azure.
+
+## Details
+
+### Discovery
+
+Support for auto-discovering GCP VMs will be added to the Discovery Service.
+
+```yaml
+discovery_service:
+  enabled: 'yes'
+  gcp:
+    - types: ['vm']
+      project_ids: ['<project-id>']
+      locations: ['<location>']
+      tags:
+        'teleport': 'yes'
+      install:
+        join_params:
+          token_name: 'gcp-discovery-token' # default value
+        nodename:
+        script_name: 'default-installer' # default value
+```
+
+New GCP nodes will be discovered periodically on a 5 minute timer, as new
+nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall Teleport on top of an instance where it is
+already present, the generated Teleport config will match against the node name using
+the project ID, zone, and instance ID by default. This can be overridden
+by specifying a node name in the join params.
+
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "<project-id>-<zone>-<instance-id>",
+    "labels": {
+      "env": "example",
+      "teleport.internal/discovered-node": "yes",
+      "teleport.internal/discovered-by": "<discovered-node-uuid>",
+      "teleport.internal/origin": "cloud",
+      "teleport.internal/zone": "<zone>",
+      "teleport.internal/projectId": "<project-id>",
+      "teleport.internal/instanceId": "<instance-id>"
+    }
+  },
+  "spec": {
+    "public_addr": "...",
+    "hostname": "gcpxyz"
+  }
+}
+```
+
+### Agent installation
+
+In order to install the Teleport agent on GCP virtual machines, Teleport will serve an
+install script at `/webapi/scripts/{installer-resource-name}`. Installer scripts will
+be editable as a resource.
+
+Example resource script:
+
+```yaml
+kind: installer
+metadata:
+  name: 'installer' # default value
+spec:
+  # shell script that will be downloaded and run by the virtual machine
+  script: |
+    #!/bin/sh
+    curl https://.../teleport-pubkey.asc ...
+    echo "deb [signed-by=... stable main" | tee ... > /dev/null
+    apt-get update
+    apt-get install teleport
+    teleport node configure --auth-agent=... --join-method=gcp --token-name=gcp-token
+  # Any resource in Teleport can automatically expire.
+  expires: 0001-01-01T00:00:00Z
+```
+
+Unless overridden by a user, a default teleport installer command will be
+generated that is appropriate for the current running version and operating
+system initially supporting DEB and RPM based distros that Teleport already
+provides packages for.
+
+To run commands on a VM, the Discovery Service will create a short-lived
```

Comment:
How long will the key be valid for and what will happen to it after we're done with installation?

Is there a way to use SSH certificate authentication with the VMs instead of key pair?

Reply:
We can set an expiry on the key in the metadata (although it won't get deleted automatically).

I don't believe there's a way to use SSH cert authentication; creating key pairs is what GCP recommends.